File: src/Python/Server/appauth.py
import hashlib
import json
import hmac

class AppIdNotFound(Exception):
    """Exception that indicates that it cannot find App ID."""    
    def __init__(self, message="App ID not found"):
        self.message = message
        super().__init__(self.message)
        

class AppAuthHeaderNotFound(Exception):
    """Exception that indicates that it cannot find Auth Header."""
    def __init__(self, message="App authentication header not found"):
        self.message = message
        super().__init__(self.message)
        

class InvalidAppAuthenticationChallenge(Exception):
    """Exception that indicates that the app authentication challenge is invalid."""
    def __init__(self, message="Invalid app authentication challenge"):
        self.message = message
        super().__init__(self.message)

class NoAppsProvided(Exception):
    """Exception that indicates that there are no authorized apps."""
    def __init__(self, message="No apps in the authorized apps list"):
        self.message = message
        super().__init__(self.message)

class InvalidBody(Exception):
    """Exception indicating that the body field must be a string or a JSON dictionary."""    
    def __init__(self, message="The body must be a string or a JSON dictionary"):
        self.message = message
        super().__init__(self.message)

class InvalidMethod(Exception):
    """Exception indicating that the method is not allowed"""
    def __init__(self, received_method, allowed_methods):
        self.message = "The method {} isn't allowed. Allowed methods are {}".format(received_method,", ".join(allowed_methods))
        super().__init__(self.message)


class AppAuthenticationServer(object):
    def __init__(self, apps):
        r"""Initializes the AppAuthenticationServer

        Parameters
        ----------
            apps : array of tuples
                The array of apps allowed to access this API, the tuple should contain the (appid, appkey)

        Raises
        ------
        NoAppsProvided
            There are no authorized apps in the list
        """
        if len(apps) <= 0:
            raise NoAppsProvided()
        self._apps = apps

    
    def authenticateApp(self, headers, method="GET", body=None, isjson=True, customsig=None):
        r"""Tries to authenticate an app with the API

        Parameters
        ----------
            headers : dictionary
                The headers of the request
            method : {'GET', 'POST', 'PATCH', 'PUT', 'HEAD', 'DELETE' }, optional
                The method of the request, can be one of the options in the brackets, defaults to the first one
            body : str or dictionary
                This variable can either be a str or a dictionary containing the body of the request
            isjson : bool
                The body is a JSON dictionary, defaults to True
            customsig : str
                A string to build a custom signature, more information in the repo wiki. Defaults to None
            
        Raises
        ------
        AppIdNotFound
            Unable to find the App Id in the authorized apps
        InvalidAppAuthenticationChallenge
            Unable to authenticate the app (invalid challenge)
        AppAuthHeaderNotFound
            Unable to find the required headers to authenticate the app
        InvalidBody
            The Body content isn't a dictionary nor a str
        InvalidMethod
            The method of the request is not valid for this headers
        """
        try:
            appid = headers["X-App-Id"]
            appkey = self.getAppKey(appid)
            timestamp = headers["X-Req-Timestamp"]
            nonce = headers["X-Req-Nonce"]
            sig = headers["X-Req-Sig"]

            if appkey is None:
                raise AppIdNotFound()

            if body is None:
                if self.compareSigWithoutBody(method, timestamp, nonce, appid, appkey, sig, customsig):
                    return True
            else:
                if self.compareSigWithBody(method, timestamp, nonce, appid, appkey, sig, body, isjson, customsig):
                    return True

            raise InvalidAppAuthenticationChallenge()
        except KeyError:
            raise AppAuthHeaderNotFound()

    def getAppKey(self, appid):
        r"""Fetches the app key given an app id from the apps list in the current instance

        Parameters
        ----------
            appid : str
                The id of he app for which we will search the key

        Returns
        -------
            str
                Returns the app key if not found returns None
        """
        for app in self._apps:
            if app[0] == appid:
                return app[1]
        return None

    def compareSigWithBody(self, method, timestamp, nonce, appid, key, hsig, body, isjson, customsig):
        r"""Compares the header signature with the computed signature for a request with body

        Parameters
        ----------
            method : {'POST', 'PATCH', 'PUT' }
                The method of the request, can be one of the options in brackets
            timestamp : str
                Timestamp of the request (sent through the headers)
            nonce : str
                Nonce of the request (sent through the headers)
            appid : str
                App trying to access the API (sent through the headers)
            key : str
                The key associated with the app id
            hsig : str
                The request signature (sent through the headers)
            body : str or dictionary
                This variable can either be a str or a dictionary containing the body of the request            
            isjson : bool, optional
                The body is a JSON dictionary, defaults to True
            customsig : str
                A string to build a custom signature, more information in the repo wiki. Defaults to None

        Returns
        -------
            bool
                True if the challenge is valid, False otherwise

        Raises
        ------
        InvalidBody
            The Body content isn't a dictionary nor a str
        InvalidMethod
            The method of the request is not valid for this headers
        """

        allowed = False
        allowed_methods = ("POST", "PATCH", "PUT")
        for m in allowed_methods:
            if m == method:
                allowed = True
                break
        if not allowed:
            raise InvalidMethod(method, allowed_methods)

        if isjson:
            bodyhash = hashlib.sha256(json.dumps(body).encode('utf-8')).hexdigest()
        else:
            if isinstance(body, str):
                bodyhash = hashlib.sha256(json.dumps(body).encode('utf-8')).hexdigest()
            else:
                raise InvalidBody()

        if customsig is None:
            sig = "{appid}{method}{timestamp}{nonce}{bodyhash}".format(
                appid = appid,
                method = method,
                timestamp = timestamp, 
                nonce = nonce,
                bodyhash = bodyhash
            )
        else:
            sig = customsig.format(
                appid = appid,
                method = method,
                timestamp = timestamp,
                nonce = nonce,
                bodyhash = bodyhash
            )

        hmacsh256 = hmac.new(
            key=key.encode('utf-8'), 
            msg=sig.encode('utf-8'), 
            digestmod=hashlib.sha256
        )

        return hmacsh256.hexdigest() == hsig


    def compareSigWithoutBody(self, method, timestamp, nonce, appid, key, hsig, customsig):
        r"""Compares the header signature with the computed signature for a request without body

        Parameters
        ----------
            method : {'GET', 'HEAD', 'DELETE' }
                The method of the request, can be one of the options in brackets
            timestamp : str
                Timestamp of the request (sent through the headers)
            nonce : str
                Nonce of the request (sent through the headers)
            appid : str
                App trying to access the API (sent through the headers)
            key : str
                The key associated with the app id
            hsig : str
                The request signature (sent through the headers)
            customsig : str
                A string to build a custom signature, more information in the repo wiki. Defaults to None

        Returns
        -------
            bool
                True if the challenge is valid, False otherwise

        Raises
        ------
        InvalidMethod
            The method of the request is not valid for this headers
        """

        allowed = False
        allowed_methods = ("GET", "HEAD", "DELETE")
        for m in allowed_methods:
            if m == method:
                allowed = True
                break
        if not allowed:
            raise InvalidMethod(method, allowed_methods)

        if customsig is None:
            sig = "{appid}{method}{timestamp}{nonce}".format(
                appid = appid,
                method = method,
                timestamp = timestamp, 
                nonce = nonce
            )
        else:
            sig = customsig.format(
                appid = appid,
                method = method,
                timestamp = timestamp,
                nonce = nonce
            )

        hmacsh256 = hmac.new(
            key=key.encode('utf-8'), 
            msg=sig.encode('utf-8'), 
            digestmod=hashlib.sha256
        )

        return hmacsh256.hexdigest() == hsig

File: src/Swift/Server/AppAuth.swift
//
// AppAuth.swift
// Authenticate an APP with the API
//
// Created by Pedro cavaleiro on 25/05/2021
// Copyright © 2021 Pedro Cavaleiro. All rights reserved.
//
import Foundation
import CommonCrypto

class AppAuthenticationServer {

    /// Errors related with the authentication of the app and api
    enum AuthenticationError: Error {
        case AppIdNotFound
        case HeadersNotFound(missingHeader: Headers)
        case InvalidChallenge
    }

    /// Internal errors not triggered by the app authentication
    enum CompareErrors: Error {
        case NoAppsProvided
    }

    /// Allowed methods with body
    enum PostMethods: String {
        case POST = "POST"
        case PATCH = "PATCH"
        case PUT = "PUT"
    }

    /// Allowed methods without body
    enum GetMethods: String {
        case GET = "GET"
        case HEAD = "HEAD"
        case DELETE = "DELETE"
    }

    /// Natively supported crypto algorithms
    private enum CryptoAlgorithm {
        case MD5, SHA1, SHA224, SHA256, SHA384, SHA512

        /// Natively supported HMAC algorithms
        var HMACAlgorithm: CCHmacAlgorithm {
            var result: Int = 0
            switch self {
            case .MD5:      result = kCCHmacAlgMD5
            case .SHA1:     result = kCCHmacAlgSHA1
            case .SHA224:   result = kCCHmacAlgSHA224
            case .SHA256:   result = kCCHmacAlgSHA256
            case .SHA384:   result = kCCHmacAlgSHA384
            case .SHA512:   result = kCCHmacAlgSHA512
            }
            return CCHmacAlgorithm(result)
        }

        /// Length for each algorithm
        var digestLength: Int {
            var result: Int32 = 0
            switch self {
            case .MD5:      result = CC_MD5_DIGEST_LENGTH
            case .SHA1:     result = CC_SHA1_DIGEST_LENGTH
            case .SHA224:   result = CC_SHA224_DIGEST_LENGTH
            case .SHA256:   result = CC_SHA256_DIGEST_LENGTH
            case .SHA384:   result = CC_SHA384_DIGEST_LENGTH
            case .SHA512:   result = CC_SHA512_DIGEST_LENGTH
            }
            return Int(result)
        }
    }

    /// Headers required for the authentication
    enum Headers: String, CaseIterable {
        case sig = "X-Req-Sig"
        case appId = "X-App-Id"
        case nonce = "X-Req-Nonce"
        case timestamp = "X-Req-Timestmap"
    }

    /// Global variable holding all the authorized apps
    private var authorizedApps: [String: String] = [String: String]()

    /**
     Initializer
     
     Initializes the AppAuthenticationServer with a list of authorized apps
     
     - parameters:
        - authorizedApps: A list of authorized apps as a dictionary containing the appId and appKey
     - Precondition: The `authorizedApps` must contain at least one element.
     - Throws:
        - CompareErrors.NoAppsProvided: When `authorizedApps` contains zero elements.
     */
    init(authorizedApps: [String: String]) throws {
        if authorizedApps.isEmpty {
            throw CompareErrors.NoAppsProvided
        }
        self.authorizedApps = authorizedApps
    }

    /**
     Authenticates the app with the api
     
     Tries to authenticate the app with the api for the methods "GET", "HEAD" and "DELETE"
     
     - parameters:
        - headers: The headers of the request, must be at least the headers mentioned on the `Headers` enum
        - method: The method of the request, optional, defaults to `.GET`
        - customSig: The signature format that will be used to generate the request signature, optional, defaults to "{appid}{method}{timestamp}{nonce}"
     - Precondition: The headers dictionary must contain all the headers mentioned on the `Headers` enum
     - Throws:
        - `AuthenticationError.HeadersNotFound(missingHeader:)`: When some header is missing, the missing header is referenced
        - `AuthenticationError.AppIdNotFound`: The appid was not found in the `authorizedApps`
     - returns: True if authenticated, no other return as all other paths lead to exceptions
     */
    func authenticateApp(headers: [String: String], method: GetMethods = .GET, customSig: String = "{appid}{method}{timestamp}{nonce}") throws -> Bool {
        try headersPresent(headers: headers)
        
        let appId = headers.first(where: { $0.key == Headers.appId.rawValue })!
        let timestamp = headers.first(where: { $0.key == Headers.timestamp.rawValue })!
        let nonce = headers.first(where: { $0.key == Headers.nonce.rawValue })!
        let sig = headers.first(where: { $0.key == Headers.sig.rawValue })!
        
        guard let appKey = authorizedApps.first(where: { $0.key == appId.value }) else { throw AuthenticationError.AppIdNotFound }
        
        var signature = customSig
        let values = [
            "timestamp": timestamp.value,
            "nonce": nonce.value,
            "appid": appId.value,
            "method": method.rawValue
        ]
        
        replacePlaceholders(signature: &signature, values: values)
        
        if hmac(signature, algorithm: .SHA256, key: appKey.value) == sig.value {
            return true
        }
        throw AuthenticationError.InvalidChallenge
    }
    
    /**
     Authenticates the app with the api
     
     Tries to authenticate the app with the api for the methods "POST", "PATCH" and "PUT"
     
     - parameters:
        - headers: The headers of the request, must be at least the headers mentioned on the `Headers` enum
        - body: The body of the request, must be a codable struct or a `String` if body is a `String` parameter `isJson` must be set to `false`
        - method: The method of the request, optional, defaults to `.POST`
        - isJson: Sets if the body is a JSON dictionary `[String: String]` or a `String`
        - customSig: The signature format that will be used to generate the request signature, optional, defaults to "{appid}{method}{timestamp}{nonce}{bodyhash}"
     - Precondition: The headers dictionary must contain all the headers mentioned on the `Headers` enum
     - Precondition: The `body` must be a codable struct or a `String`, when the type is `String` the parameter `isJson` must be set to `false`
     - Throws:
        - `AuthenticationError.HeadersNotFound(missingHeader:)`: When some header is missing, the missing header is referenced
        - `AuthenticationError.AppIdNotFound`: The appid was not found in the `authorizedApps`
     - returns: True if authenticated, no other return as all other paths lead to exceptions
     */
    func authenticateApp<T: Codable>(headers: [String: String], body: T, method: PostMethods = .POST, isJson: Bool = true, customSig: String = "{appid}{method}{timestamp}{nonce}{bodyhash}") throws -> Bool {
        try headersPresent(headers: headers)
        
        let appId = headers.first(where: { $0.key == Headers.appId.rawValue })!
        let timestamp = headers.first(where: { $0.key == Headers.timestamp.rawValue })!
        let nonce = headers.first(where: { $0.key == Headers.nonce.rawValue })!
        let sig = headers.first(where: { $0.key == Headers.sig.rawValue })!
        
        guard let appKey = authorizedApps.first(where: { $0.key == appId.value }) else { throw AuthenticationError.AppIdNotFound }
        
        var signature = customSig
        let values = [
            "timestamp": timestamp.value,
            "nonce": nonce.value,
            "appid": appId.value,
            "method": method.rawValue,
            "bodyhash": isJson ? sha256(value: String(data: (try! JSONEncoder().encode(body)), encoding: .utf8)!) : sha256(value: (body as! String))
        ]
        
        replacePlaceholders(signature: &signature, values: values)
        
        if hmac(signature, algorithm: .SHA256, key: appKey.value) == sig.value {
            return true
        }
        throw AuthenticationError.InvalidChallenge
    }

    /**
     Checks if required headers are present
     
     - parameters:
        - headers: A dictionary of the request headers
     - Throws
        - `AuthenticationError.HeadersNotFound(missingHeader: header)`: When there's a missing header
     */
    private func headersPresent(headers: [String: String]) throws {
        for header in Headers.AllCases() {
            guard (headers.first(where: { $0.key == header.rawValue }) != nil) else { throw AuthenticationError.HeadersNotFound(missingHeader: header) }
        }
    }

    /**
     Creates the SHA256 hash of the string and returns it in hex format as string
     
     - parameters:
        - value: The string from which will be generated the hash
     - returns: The hash in hex format as string
     */
    private func sha256(value: String) -> String {
        let data = Data(value.utf8)
        var hash = [UInt8](repeating: 0,  count: Int(CC_SHA256_DIGEST_LENGTH))
        data.withUnsafeBytes {
            _ = CC_SHA256($0.baseAddress, CC_LONG(data.count), &hash)
        }
        return hash.map { String(format: "%02x", $0) }.joined()
    }

    /**
     Creates the HMAC of the string and returns it in hex format as string
     
     - parameters:
        - value: The string from which will be generated the HMAC
        - algorithm: The algorithm of the HMAC, reffer to `CryptoAlgorithm` for the supported algorithms
        - key: The key as string to generate the HMAC
     - returns: The HMAC in hex format as string
     */
    private func hmac(_ value: String, algorithm: CryptoAlgorithm, key: String) -> String {
        let str = value.cString(using: String.Encoding.utf8)
        let strLen = Int(value.lengthOfBytes(using: String.Encoding.utf8))
        let digestLen = algorithm.digestLength
        let result = UnsafeMutablePointer<CUnsignedChar>.allocate(capacity: digestLen)
        let keyStr = key.cString(using: String.Encoding.utf8)
        let keyLen = Int(key.lengthOfBytes(using: String.Encoding.utf8))

        CCHmac(algorithm.HMACAlgorithm, keyStr!, keyLen, str!, strLen, result)

        let digest = stringFromResult(result: result, length: digestLen)

        result.deallocate()

        return digest
    }

    /**
     Converts the HMAC given in `UnsafeMutablePointer<CUnsignedChar>` to a hex string
     
     - parameters:
        - result: The value of the HMAC
        - length: The length of the HMAC
     - returns: The HMAC in hex format as string
     */
    private func stringFromResult(result: UnsafeMutablePointer<CUnsignedChar>, length: Int) -> String {
        let hash = NSMutableString()
        for i in 0..<length {
            hash.appendFormat("%02x", result[i])
        }
        return String(hash).lowercased()
    }

    /**
     Replaces the placeholders with the propper values
     
     The replacement of the placeholders with the propper values is done "in place" thus not requiring a return value
     
     - parameters:
        - signature: The string with the placeholders
        - values: A dictionary with the values to replace
     */
    private func replacePlaceholders(signature: inout String, values: [String: String]) {
        for value in values {
            signature = signature.replacingOccurrences(of: "{\(value.key)}", with: value.value)
        }
    }

}
